### PR TITLE
Post/java stream collect pitfalls

### PR DESCRIPTION
**Pull request by nfaupel**

### CODE REVIEW

File: _posts/2018-07-05-java-stream-collect-pitfalls.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:502)
+    at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
+    at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
+    at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
+    ...
+```
+
+The merge method of the `HashMap` used by the `Collectors#toMap()` does not allow values to be `null` although the `HashMap` allows `null` values in general to be stored in the map.
+
+So it depends on your use case if you can rewrite every for-loop to a stream collect based code style and if you are really sure that you will never have `null` values.
```
Comment (hilton) [inline _posts/2018-07-05-java-stream-collect-pitfalls.md]: This sentence is awkward; I suggest:

> Whether you can rewrite every for-loop to a stream collect-based code style depends on your use case. You also need to be really sure that you will never have `null` values.

File: _posts/2018-07-05-java-stream-collect-pitfalls.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+Having that in mind and that the documentation does not always describe the actual behavior of the code in every edge case you should be sure what used libraries are doing and how they handle edge cases.
```
Comment (hilton) [inline _posts/2018-07-05-java-stream-collect-pitfalls.md]: Change 'used libraries' to 'the libraries you use'

File: _posts/2018-07-05-java-stream-collect-pitfalls.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+And as always it is a good idea to test your code with data that is as close to your productive data as possible including edge cases.
```
Comment (hilton) [inline _posts/2018-07-05-java-stream-collect-pitfalls.md]: Add commas:

> And, as always, it is…

Comment (hilton) [inline _posts/2018-07-05-java-stream-collect-pitfalls.md]: Change 'productive' to 'production' (classic German-English)

File: _posts/2018-07-05-java-stream-collect-pitfalls.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
+---
+title: "Java 8 stream collect to Map pitfalls"
```
Comment (hilton) [inline _posts/2018-07-05-java-stream-collect-pitfalls.md]: The title is hard to read because 'collect to Map' is really Java, not English. How about 'Trouble building maps with the Java Stream API'?

File: _posts/2018-07-05-java-stream-collect-pitfalls.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+description: "Pitfalls when rewriting for-loops with java streams and collect to Map"
+tags: java, java8, stream, for, loop, map, hashmap, collect
+author: Nils Faupel
+layout: article
+---
+
+# Adding items from a list to a map
+
+Let's assume you have a `List` of `Person`s with ID and name and you want to create a `Map` from ID to the name.
```
Comment (hilton) [inline _posts/2018-07-05-java-stream-collect-pitfalls.md]: Please use ‘curly’ quotes

File: _posts/2018-07-05-java-stream-collect-pitfalls.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:512)
+    at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:502)
+    at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
+    at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
+    at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
+    ...
+```
+
+So where does this Exception come from?
+If you look into the documentation [Collectos#toMap()] (https://docs.oracle.com/javase/8/docs/api/java/util/stream/Collectors.html#toMap-java.util.function.Function-java.util.function.Function-) you will find the following hint:
```
Comment (hilton) [inline _posts/2018-07-05-java-stream-collect-pitfalls.md]: The Markdown link is broken

**Comment by nfaupel** [comment]
Applied the requested changes from Peter.
